sources/packages/backend/libs/services/src/students-assessments/assessment-sequential-processing.service.ts

```diff
@@ -0,0 +1,182 @@
+    // If all the COEs from the assessment are declined some user action will be needed in the application
+    // and this application will not be considered for sequential processing.
+    const existsValidCOE = entityManager
+      .getRepository(DisbursementSchedule)
+      .createQueryBuilder("disbursementSchedule")
+      .select("1")
+      .where("disbursementSchedule.studentAssessment.id = currentAssessment.id")
+      .andWhere("disbursementSchedule.coeStatus != :declinedCOEStatus")
+      .limit(1)
+      .getSql();
+    // Returns past, current, and future applications ordered by the first ever executed assessment calculation.
+    const sequentialApplications = await entityManager
+      .getRepository(Application)
+      .createQueryBuilder("application")
+      .select("application.id", "applicationId")
+      .addSelect("application.applicationNumber", "applicationNumber")
+      .addSelect("application.applicationStatus", "applicationStatus")
+      .addSelect("currentAssessmentOffering.id", "currentAssessmentOfferingId")
+      .addSelect(`(${assessmentDateSubQuery})`, "referenceAssessmentDate")
+      .innerJoin("application.student", "student")
+      .innerJoin("application.programYear", "programYear")
+      .innerJoin("application.currentAssessment", "currentAssessment")
+      .innerJoin("currentAssessment.offering", "currentAssessmentOffering")
+      .where("student.id = :studentId", { studentId })
+      .andWhere("programYear.id = :programYearId", { programYearId })
+      .andWhere("application.applicationStatus != :overwrittenStatus", {
+        overwrittenStatus: ApplicationStatus.Overwritten,
+      })
+      .andWhere(
+        new Brackets((qb) => {
+          qb.where(
+            // Conditions to find applications considered before the current application and after.
+            // 1. Application should not be cancelled;
+            // 2. Current assessment must have a date would indicates that it was calculated.
+            // 3. At least one COE must be valid.
+            new Brackets((pastOrFutureBracket) =>
+              pastOrFutureBracket
+                .where("application.applicationStatus != :cancelledStatus", {
+                  cancelledStatus: ApplicationStatus.Cancelled,
+                })
+                .andWhere("currentAssessment.assessmentDate IS NOT NULL")
+                .andWhere(`EXISTS (${existsValidCOE})`),
+            ),
+            // The 'or' condition forces the current application to be returned since its data matters for decisions,
+            // for instance, if the current application has no calculation date no future application should be impacted.
+          ).orWhere("application.applicationNumber = :applicationNumber", {
+            applicationNumber,
+          });
+        }),
+      )
+      .setParameters({ declinedCOEStatus: COEStatus.declined })
+      .orderBy(`"referenceAssessmentDate"`)
```

Comment:
optional. referenceAssessmentDate can be const

Reply:
Audit fields added.